pvlib/pvsystem.py

```diff
@@ -1647,26 +1647,39 @@ def retrieve_sam(name=None, path=None):
     return _parse_raw_sam_df(csvdata)
 
 
+def _normalize_sam_product_names(names):
+    '''
+    Replace special characters within the product names to make them more
+    suitable fur use as Dataframe column names.
+    '''
+    import warnings
+
+    BAD_CHARS  = ' -.()[]:+/",'
+    GOOD_CHARS = '____________'
+
+    mapping = str.maketrans(BAD_CHARS, GOOD_CHARS)
+    names = pd.Series(data=names)
+    norm_names = names.str.translate(mapping)
+
+    n_duplicates = names.duplicated().sum()
+    if n_duplicates > 0:
+        warnings.warn ('Original names contain %d duplicate(s).' % n_duplicates)
+
+    n_duplicates = norm_names.duplicated().sum()
+    if n_duplicates > 0:
```

Comment:
are either of these warnings triggered in the current file? I would have thought the DataFrame construction would fail if so.

Reply:
I got 6 duplicates after normalization in CECmod, and 12 before normalization in ADRInverter.  It is possible to have duplicate values in DataFrame columns or index.

Reply:
I can add a test if the function gets the thumbs up.

Reply:
Do I see correctly that there is no current test for `retrieve_sam()`?

Reply:
@janinefreeman FYI. I recognize that both SAM and pvlib are victims of what appears in the CEC table. Currently, pvlib imports the module parameter file from SAM. Perhaps there's a way we can help reduce some of these formatting requirements upstream in SAM, and avoid the duplicates.

Reply:
@cwhanse Thanks for the heads up. We get frequent requests to fix info coming from the CEC, but we are trying very hard to avoid error-checking the database. But if this method could be ported into SAM, that could be a possibility. Yet another scenario where we'd all benefit from the oft-pitched crowd-sourced component database...

Reply:
@janinefreeman I think this is a case of creating code-friendly row and column labels. If you can point me to the part of SAM code that reads the CEC's spreadsheet, or that writes the SAM module database, I'm happy to take a look.